{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulation\n",
    "\n",
    "* make 4 sets of sequences: 250 seq , 1000 seqs, 5000 seqs, each 250 bp (meme takes 60000 only)\n",
    "    * Use random generator with real frequencies\n",
    "* Insert real motifs but have been modified to have mC.\n",
    "    * CEBPB , NRF1, SP1, CTCF, made up motif\n",
    "* Different abundances: 1%, 2%, 5%, 10%, 25% of peaks\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os \n",
    "from sys import argv\n",
    "import random as rd\n",
    "from copy import deepcopy\n",
    "\n",
    "def load_motifsEF(filename):\n",
    "    file=open(filename)\n",
    "    seq=file.read().split(\"MOTIF\")\n",
    "    seq=seq[1:]\n",
    "    motifs={}\n",
    "    infos={}\n",
    "    for s in range(len(seq)):\n",
    "        t=seq[s].strip().split(\"\\n\")\n",
    "        motifs[t[0]]=t[2:]  #remove HOCOMOCO URL\n",
    "        infos[t[0].split('_')[0]]=t[:2]\n",
    "    for m in motifs:\n",
    "        tdict={'A':[],'C':[],'G':[],'T':[],'E':[],'F':[]}\n",
    "        for pos in range(len(motifs[m])):\n",
    "            tmp=motifs[m][pos].strip().split(\"\\t\")\n",
    "            tdict['A']+=[float(tmp[0])]\n",
    "            tdict['C']+=[float(tmp[1])]\n",
    "            tdict['G']+=[float(tmp[2])]\n",
    "            tdict['T']+=[float(tmp[3])]\n",
    "            tdict['E']+=[float(tmp[4])]\n",
    "            tdict['F']+=[float(tmp[5])]\n",
    "        motifs[m]=tdict\n",
    "    \n",
    "    return motifs,infos\n",
    "\n",
    "\n",
    "def load_motifs(filename):\n",
    "    file=open(filename)\n",
    "    seq=file.read().split(\"MOTIF\")\n",
    "    seq=seq[1:]\n",
    "    motifs={}\n",
    "    infos={}\n",
    "    for s in range(len(seq)):\n",
    "        t=seq[s].strip().split(\"\\n\")\n",
    "        motifs[t[0]]=t[2:-1]  #remove HOCOMOCO URL\n",
    "        infos[t[0].split('_')[0]]=t[:2]\n",
    "    for m in motifs:\n",
    "        tdict={'A':[],'C':[],'G':[],'T':[]}\n",
    "        for pos in range(len(motifs[m])):\n",
    "            tmp=motifs[m][pos].strip().split(\"\\t\")\n",
    "            tdict['A']+=[float(tmp[0])]\n",
    "            tdict['C']+=[float(tmp[1])]\n",
    "            tdict['G']+=[float(tmp[2])]\n",
    "            tdict['T']+=[float(tmp[3])]\n",
    "            #tdict['E']+=[float(tmp[4])]\n",
    "            #tdict['F']+=[float(tmp[5])]\n",
    "        motifs[m]=tdict\n",
    "    \n",
    "    return motifs,infos\n",
    "def insertBaseToPWM(pwm,loc,prop):\n",
    "    #randomly insert modified bases to an existing pwm\n",
    "    #insert at loc location, probability is prop\n",
    "    newpwm = deepcopy(pwm)\n",
    "    k = len(pwm['A'])\n",
    "    base = rd.choice(['E','E']) # only adds E, for this is typeE\n",
    "    freq = prop #rd.choice(range(5,10))*0.1 #prop 0.5 - 1.0\n",
    "    print base,loc,freq\n",
    "    newpwm['E'] = [0]*k #add other empty columns to the matrix\n",
    "    newpwm['F'] = [0]*k\n",
    "\n",
    "    for c in newpwm.keys():\n",
    "        newpwm[c][loc] = (1-freq)/(len(newpwm.keys())-1)\n",
    "    newpwm[base][loc] = freq\n",
    "    return newpwm\n",
    "\n",
    "def insertBaseToPWM_typeE(pwm,loc,props):\n",
    "    #randomly insert modified bases to an existing pwm\n",
    "    #insert at loc location, probability is prop\n",
    "    newpwm = deepcopy(pwm)\n",
    "    k = len(pwm['A'])\n",
    "    newpwm['E'] = [0]*k #add another empty column to the matrix\n",
    "    \n",
    "    for c in newpwm.keys():\n",
    "        newpwm[c][loc] = props[c]\n",
    "    return newpwm\n",
    "\n",
    "def randomkmer_typeEF(PWM):\n",
    "    #take a PWM and randomly create a Kmer based on it\n",
    "    # chance goes A->C->G->T->E->F\n",
    "    kmer=\"\"\n",
    "    for pos in range(len(PWM['A'])):\n",
    "        die=rd.randint(0,100000)\n",
    "        if die<=PWM['A'][pos]*100000:\n",
    "            kmer+='A'\n",
    "        elif die<=(PWM['A'][pos]+PWM['C'][pos])*100000:\n",
    "            kmer+='C'\n",
    "        elif die<=(PWM['A'][pos]+PWM['C'][pos]+PWM['G'][pos])*100000:\n",
    "            kmer+='G'\n",
    "        elif die<=(PWM['A'][pos]+PWM['C'][pos]+PWM['G'][pos]+PWM['T'][pos])*100000:\n",
    "            kmer+='T'\n",
    "        elif die<=(PWM['A'][pos]+PWM['C'][pos]+PWM['G'][pos]+PWM['T'][pos]+PWM['E'][pos])*100000:\n",
    "            kmer+='E'\n",
    "        else:\n",
    "            kmer+='F'\n",
    "    return kmer\n",
    "\n",
    "def loadFasta(filename):\n",
    "    seqs = {}\n",
    "    lines = open(filename).read().split('>')[1:]\n",
    "    for line in lines:\n",
    "        tmp = line.split('\\n')\n",
    "        seqs[tmp[0]] = ''.join(tmp[1:])\n",
    "    return seqs\n",
    "\n",
    "def insertoseqs(pwm,seqs,numseqtopick):\n",
    "    #take pwm, the seqs, and insert the pwm to numseqtopick of the seqs \n",
    "    \n",
    "    seqspicked = rd.sample(seqs.keys(),numseqtopick)\n",
    "    #print seqspicked\n",
    "    for s in seqspicked:\n",
    "        kmer = randomkmer_typeEF(pwm)\n",
    "        #print kmer\n",
    "        string = list(seqs[s])\n",
    "        insertloc = rd.randrange(0,len(string)-len(kmer))\n",
    "        #print insertloc\n",
    "        for i in xrange(len(kmer)):\n",
    "            string[insertloc + i]=kmer[i]\n",
    "    \n",
    "        seqs[s] = ''.join(string)\n",
    "    return seqs\n",
    "def writeoutPWM(pwm,name,outfile):\n",
    "    target = open(outfile,'w')\n",
    "    target.write(\"MOTIF\\t\"+name+'\\n')\n",
    "    for loc in range(len(pwm['A'])):\n",
    "        line = []\n",
    "        for base in ['A','C','G','T','E','F']:\n",
    "            if base not in pwm: continue\n",
    "            line += [str(pwm[base][loc])]\n",
    "        line = '\\t'.join(line)\n",
    "        target.write(line+'\\n')\n",
    "    target.close()\n",
    "    return\n",
    "def printPWM(pwm):\n",
    "    print '\\t'.join(['A','C','G','T','E','F'])\n",
    "    for loc in range(len(pwm['A'])):\n",
    "        line = []\n",
    "        for base in ['A','C','G','T','E','F']:\n",
    "            if base in pwm:\n",
    "                line += [str(round(pwm[base][loc],3))]\n",
    "        print '\\t'.join(line)\n",
    "    return\n",
    "def getConsensus(pwm):\n",
    "    #take the dict pwm and print the consensus of the pwm\n",
    "    cnss = '' \n",
    "    for loc in range(len(pwm['A'])):\n",
    "        maxbase = 'A'\n",
    "        maxvalue = 0.0\n",
    "        for base in pwm:\n",
    "            if pwm[base][loc] > maxvalue:\n",
    "                maxbase = base\n",
    "                maxvalue = pwm[base][loc]\n",
    "        cnss += maxbase\n",
    "    return cnss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "knownMotifs,knownInfos = load_motifs(\"HOCOMOCOv11_core_HUMAN_mono_meme_format.meme\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTCF_HUMAN.H11MO.0.A\n",
      "TGGCCACCAGGGGGCGCCA\n",
      "CEBPB_HUMAN.H11MO.0.A\n",
      "GATTGTGCAATA\n",
      "NRF1_HUMAN.H11MO.0.A\n",
      "CACTGCGCATGCGCAGG\n",
      "GATAEGCA\n"
     ]
    }
   ],
   "source": [
    "mymotifs = {}\n",
    "for key in knownMotifs:\n",
    "    if \"CTCF_\" in key or \"CEBPB\" in key or \"NRF1\" in key:\n",
    "        print key\n",
    "        #print knownMotifs[key]\n",
    "        print getConsensus(knownMotifs[key])\n",
    "        mymotifs[key.split('_')[0]] = knownMotifs[key]\n",
    "        \n",
    "\n",
    "#add the methylation to these motifs\n",
    "#CEBPB: no. 5\n",
    "props = {\n",
    "    'A':0.024376,\n",
    "    'C':0.085981,\n",
    "    'G':0.014709,\n",
    "    'T':0.214204,\n",
    "    'E':0.660729\n",
    "        }\n",
    "loc = 5 \n",
    "mymotifs['CEBPB'] = insertBaseToPWM_typeE(loc=loc, props=props, pwm=mymotifs['CEBPB'])\n",
    "\n",
    "#CTCF: loc = 11\n",
    "loc = 11\n",
    "props = {\n",
    "    'A':0.0,\n",
    "    'C':0.2,\n",
    "    'G':0.0,\n",
    "    'T':0.0,\n",
    "    'E':0.8\n",
    "        }\n",
    "\n",
    "mymotifs['CTCF'] = insertBaseToPWM_typeE(loc=loc, props=props, pwm=mymotifs['CTCF'])\n",
    "\n",
    "#NRF1: loc = 11\n",
    "loc = 11\n",
    "props = {\n",
    "    'A':0.0,\n",
    "    'C':0.2,\n",
    "    'G':0.0,\n",
    "    'T':0.0,\n",
    "    'E':0.8\n",
    "        }\n",
    "\n",
    "mymotifs['NRF1'] = insertBaseToPWM_typeE(loc=loc, props=props, pwm=mymotifs['NRF1'])\n",
    "#add the gatEGaca motif too \n",
    "gataEGca = {\n",
    "    'A':[0.1,0.7,0.0,0.8,0.0,0.0,0.1,0.9],\n",
    "    'C':[0.0,0.1,0.1,0.2,0.2,0.0,0.9,0.0],\n",
    "    'G':[0.9,0.0,0.0,0.0,0.0,0.8,0.0,0.1],\n",
    "    'T':[0.0,0.1,0.9,0.0,0.0,0.1,0.0,0.0],\n",
    "    'E':[0.0,0.1,0.0,0.0,0.8,0.1,0.0,0.0],\n",
    "}\n",
    "print getConsensus(gataEGca)\n",
    "mymotifs['gataEGca'] = gataEGca\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CEBPB GATTGEGCAATA\n",
      "A\tC\tG\tT\tE\tF\n",
      "0.238\t0.099\t0.352\t0.311\t0.0\n",
      "0.454\t0.132\t0.388\t0.026\t0.0\n",
      "0.0\t0.002\t0.002\t0.996\t0.0\n",
      "0.0\t0.0\t0.3\t0.7\t0.0\n",
      "0.302\t0.002\t0.513\t0.183\t0.0\n",
      "0.024\t0.086\t0.015\t0.214\t0.661\n",
      "0.0\t0.0\t1.0\t0.0\t0.0\n",
      "0.081\t0.899\t0.0\t0.02\t0.0\n",
      "1.0\t0.0\t0.0\t0.0\t0.0\n",
      "1.0\t0.0\t0.0\t0.0\t0.0\n",
      "0.0\t0.244\t0.148\t0.608\t0.0\n",
      "0.37\t0.333\t0.183\t0.115\t0.0\n",
      "gataEGca GATAEGCA\n",
      "A\tC\tG\tT\tE\tF\n",
      "0.1\t0.0\t0.9\t0.0\t0.0\n",
      "0.7\t0.1\t0.0\t0.1\t0.1\n",
      "0.0\t0.1\t0.0\t0.9\t0.0\n",
      "0.8\t0.2\t0.0\t0.0\t0.0\n",
      "0.0\t0.2\t0.0\t0.0\t0.8\n",
      "0.0\t0.0\t0.8\t0.1\t0.1\n",
      "0.1\t0.9\t0.0\t0.0\t0.0\n",
      "0.9\t0.0\t0.1\t0.0\t0.0\n",
      "CTCF TGGCCACCAGGEGGCGCCA\n",
      "A\tC\tG\tT\tE\tF\n",
      "0.094\t0.36\t0.126\t0.42\t0.0\n",
      "0.152\t0.168\t0.522\t0.158\t0.0\n",
      "0.258\t0.084\t0.554\t0.104\t0.0\n",
      "0.068\t0.856\t0.036\t0.04\t0.0\n",
      "0.002\t0.996\t0.002\t0.0\t0.0\n",
      "0.7\t0.006\t0.242\t0.052\t0.0\n",
      "0.016\t0.652\t0.322\t0.01\t0.0\n",
      "0.098\t0.594\t0.074\t0.234\t0.0\n",
      "0.906\t0.018\t0.054\t0.022\t0.0\n",
      "0.002\t0.002\t0.992\t0.004\t0.0\n",
      "0.386\t0.0\t0.612\t0.002\t0.0\n",
      "0.0\t0.2\t0.0\t0.0\t0.8\n",
      "0.012\t0.002\t0.984\t0.002\t0.0\n",
      "0.05\t0.026\t0.86\t0.064\t0.0\n",
      "0.068\t0.876\t0.012\t0.044\t0.0\n",
      "0.396\t0.014\t0.584\t0.006\t0.0\n",
      "0.076\t0.562\t0.326\t0.036\t0.0\n",
      "0.096\t0.422\t0.116\t0.366\t0.0\n",
      "0.372\t0.268\t0.316\t0.044\t0.0\n",
      "NRF1 CACTGCGCATGEGCAGG\n",
      "A\tC\tG\tT\tE\tF\n",
      "0.108\t0.558\t0.146\t0.188\t0.0\n",
      "0.428\t0.124\t0.158\t0.29\t0.0\n",
      "0.054\t0.514\t0.342\t0.09\t0.0\n",
      "0.056\t0.1\t0.046\t0.798\t0.0\n",
      "0.09\t0.002\t0.908\t0.0\t0.0\n",
      "0.028\t0.968\t0.004\t0.0\t0.0\n",
      "0.01\t0.0\t0.988\t0.002\t0.0\n",
      "0.002\t0.992\t0.002\t0.004\t0.0\n",
      "0.782\t0.138\t0.048\t0.032\t0.0\n",
      "0.04\t0.076\t0.164\t0.72\t0.0\n",
      "0.008\t0.01\t0.976\t0.006\t0.0\n",
      "0.0\t0.2\t0.0\t0.0\t0.8\n",
      "0.0\t0.042\t0.866\t0.092\t0.0\n",
      "0.008\t0.8\t0.014\t0.178\t0.0\n",
      "0.426\t0.194\t0.316\t0.064\t0.0\n",
      "0.28\t0.14\t0.378\t0.202\t0.0\n",
      "0.114\t0.296\t0.522\t0.068\t0.0\n"
     ]
    }
   ],
   "source": [
    "for key in mymotifs:\n",
    "    print key,getConsensus(mymotifs[key])\n",
    "    printPWM(mymotifs[key])\n",
    "     "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make sequences randomly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import random\n",
    "import itertools\n",
    "import bisect\n",
    "import numpy as np\n",
    "\n",
    "def randomseqDimer(props,length):\n",
    "    \"\"\"\n",
    "    Generate string randomly at certain length\n",
    "    - props: a dict of dimers probability for DNA \n",
    "    \n",
    "    \"\"\"\n",
    "    #choices, weights = fprob_table.items()\n",
    "    fprob_table = props\n",
    "    K = length\n",
    "    choices = []\n",
    "    weights = []\n",
    "    for key in fprob_table:\n",
    "        choices += [key]\n",
    "        weights += [fprob_table[key]]\n",
    "    #cumdist = list(itertools.accumulate(weights))\n",
    "    cumdist = []\n",
    "    cumprop = 0.0\n",
    "    for i in range(len(choices)):\n",
    "        cumprop += weights[i]\n",
    "        cumdist += [cumprop]\n",
    "    results = []\n",
    "    s = \"\"\n",
    "    while len(s) < K:\n",
    "        x = random.random() * cumdist[-1]\n",
    "        s += choices[bisect.bisect(cumdist, x)]\n",
    "        #print s\n",
    "\n",
    "    return s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "### getting the dimer props \n",
    "\n",
    "backgroundfile = \"../test_data_typeE/background_typeE-5.tsv\"\n",
    "lines = open(backgroundfile).readlines()\n",
    "\n",
    "dimercount = {}\n",
    "for line in lines[1:]:\n",
    "    #print line.strip()\n",
    "    tmp = line.strip().split()\n",
    "    seq = tmp[0]\n",
    "    k = 2 \n",
    "    multiplier = int(tmp[1])\n",
    "    for i in range(len(seq) - k + 1):\n",
    "        dimer = seq[i:i+k]\n",
    "        try: \n",
    "            dimercount[dimer] += multiplier*1\n",
    "        except KeyError:\n",
    "            dimercount[dimer] = multiplier*1 \n",
    "dimerProps = {}\n",
    "total = 0.0\n",
    "for key in dimercount:\n",
    "    total += dimercount[key]\n",
    "for key in dimercount:\n",
    "    dimerProps[key] = dimercount[key]/total\n",
    "#dimerProps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CEBPB\n",
      "0.01_250\n",
      "3\n",
      "0.01_1000\n",
      "11\n",
      "0.01_5000\n",
      "51\n",
      "0.02_250\n",
      "6\n",
      "0.02_1000\n",
      "21\n",
      "0.02_5000\n",
      "101\n",
      "0.05_250\n",
      "13\n",
      "0.05_1000\n",
      "51\n",
      "0.05_5000\n",
      "251\n",
      "0.1_250\n",
      "26\n",
      "0.1_1000\n",
      "101\n",
      "0.1_5000\n",
      "501\n",
      "0.25_250\n",
      "63\n",
      "0.25_1000\n",
      "251\n",
      "0.25_5000\n",
      "1251\n",
      "gataEGca\n",
      "0.01_250\n",
      "3\n",
      "0.01_1000\n",
      "11\n",
      "0.01_5000\n",
      "51\n",
      "0.02_250\n",
      "6\n",
      "0.02_1000\n",
      "21\n",
      "0.02_5000\n",
      "101\n",
      "0.05_250\n",
      "13\n",
      "0.05_1000\n",
      "51\n",
      "0.05_5000\n",
      "251\n",
      "0.1_250\n",
      "26\n",
      "0.1_1000\n",
      "101\n",
      "0.1_5000\n",
      "501\n",
      "0.25_250\n",
      "63\n",
      "0.25_1000\n",
      "251\n",
      "0.25_5000\n",
      "1251\n",
      "CTCF\n",
      "0.01_250\n",
      "3\n",
      "0.01_1000\n",
      "11\n",
      "0.01_5000\n",
      "51\n",
      "0.02_250\n",
      "6\n",
      "0.02_1000\n",
      "21\n",
      "0.02_5000\n",
      "101\n",
      "0.05_250\n",
      "13\n",
      "0.05_1000\n",
      "51\n",
      "0.05_5000\n",
      "251\n",
      "0.1_250\n",
      "26\n",
      "0.1_1000\n",
      "101\n",
      "0.1_5000\n",
      "501\n",
      "0.25_250\n",
      "63\n",
      "0.25_1000\n",
      "251\n",
      "0.25_5000\n",
      "1251\n",
      "NRF1\n",
      "0.01_250\n",
      "3\n",
      "0.01_1000\n",
      "11\n",
      "0.01_5000\n",
      "51\n",
      "0.02_250\n",
      "6\n",
      "0.02_1000\n",
      "21\n",
      "0.02_5000\n",
      "101\n",
      "0.05_250\n",
      "13\n",
      "0.05_1000\n",
      "51\n",
      "0.05_5000\n",
      "251\n",
      "0.1_250\n",
      "26\n",
      "0.1_1000\n",
      "101\n",
      "0.1_5000\n",
      "501\n",
      "0.25_250\n",
      "63\n",
      "0.25_1000\n",
      "251\n",
      "0.25_5000\n",
      "1251\n"
     ]
    }
   ],
   "source": [
    "abundances = [0.01,0.02,0.05,0.1,0.25]\n",
    "sizes = [250, 1000, 5000]\n",
    "seq_length_avg = 250\n",
    "seq_length_std = 50\n",
    "\n",
    "resultdir = './files/SIM_seqs/'\n",
    "for motif in mymotifs:\n",
    "    print motif\n",
    "    seqsets = {} \n",
    "    for a in abundances:\n",
    "        for siz in sizes:\n",
    "            name = str(a) + '_' + str(siz)\n",
    "            print name\n",
    "            outfile = open(resultdir + motif+'_'+name+'.faa','w')\n",
    "            \n",
    "            seq_set = {}\n",
    "            for i in range(siz):\n",
    "                seqlength = int(np.random.normal(scale=seq_length_std, loc=seq_length_avg))\n",
    "                seq = randomseqDimer(props=dimerProps,length=seqlength)\n",
    "                seq_set[i] = seq \n",
    "            # now , insert the motif to the seqs\n",
    "            numtoinsert = int(a*siz) + 1\n",
    "            print numtoinsert\n",
    "            seq_set = insertoseqs(numseqtopick=numtoinsert,pwm=mymotifs['gataEGca'],seqs=seq_set)\n",
    "            \n",
    "            #write out the fasta file \n",
    "            for key in seq_set:\n",
    "                line = \">\"+str(key)+'\\n'+seq_set[key]\n",
    "                outfile.write(line+'\\n')\n",
    "            outfile.close()\n",
    "        \n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CEBPB\n",
      "gataEGca\n",
      "CTCF\n",
      "NRF1\n"
     ]
    }
   ],
   "source": [
    "for motif in mymotifs:\n",
    "    print motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for m in mymotifs:\n",
    "    writeoutPWM(name='1',outfile=m+\".SIM.typeE.meme\",pwm=mymotifs[m])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
